xdsl/dialects/riscv.py

```diff
@@ -2711,6 +2717,168 @@ def assembly_line_args(self) -> tuple[AssemblyInstructionArg, ...]:
+            },
+        )
+
+    @property
+    def max_inst(self) -> int:
+        """
+        Number of instructions to be repeated.
+        """
+        return len([op for op in self.body.ops if isinstance(op, RISCVInstruction)])
+
+    def assembly_line_args(self) -> tuple[AssemblyInstructionArg | None, ...]:
+        return (
+            self.max_rep,
+            self.max_inst,
+            self.stagger_mask.data,
+            self.stagger_count.data,
+        )
+
+    def custom_print_attributes(self, printer: Printer):
+        printer.print(", ")
+        printer.print(self.stagger_mask.data)
+        printer.print_string(", ")
+        printer.print(self.stagger_count.data)
+        return {"stagger_mask", "stagger_count"}
+
+    @classmethod
+    def custom_parse_attributes(cls, parser: Parser):
+        attributes = dict[str, Attribute]()
+        attributes["stagger_mask"] = IntAttr(
+            parser.parse_integer(
+                allow_boolean=False, context_msg="Expected stagger mask"
+            )
+        )
+        parser.parse_punctuation(",")
+        attributes["stagger_count"] = IntAttr(
+            parser.parse_integer(
+                allow_boolean=False, context_msg="Expected stagger count"
+            )
+        )
+        return attributes
+
+    def verify_(self) -> None:
+        if self.stagger_count.data:
+            raise VerifyException("Non-zero stagger count currently unsupported")
+        if self.stagger_mask.data:
+            raise VerifyException("Non-zero stagger mask currently unsupported")
+        for instruction in self.body.ops:
+            if not instruction.has_trait(Pure):
+                raise VerifyException(
+                    "Frep operation body may not contain instructions "
+                    f"with side-effects, found {instruction.name}"
+                )
+
+
+@irdl_op_definition
+class FrepOuter(FRepOperation):
+    """
+    Repeats the instruction in the body as if the body were the body of a for loop, for
+    example:
+
+    ```
+    # Repeat 4 times, stagger 1, period 2
+    li a0, 4
+    frep.outer a0, 2, 1, 0b1010
+    fadd.d fa0, ft0, ft2
+    fmul.d fa0, ft3, fa0
+    ```
+
+    is equivalent to:
+    ```
+    fadd.d fa0, ft0, ft2
+    fmul.d fa0, ft3, fa0
+    fadd.d fa1, ft0, ft3
+    fmul.d fa1, ft3, fa1
+    fadd.d fa0, ft0, ft2
+    fmul.d fa0, ft3, fa0
+    fadd.d fa1, ft0, ft3
+    fmul.d fa1, ft3, fa1
+    ```
+    """
+
+    name = "riscv.frep_outer"
+
+    def assembly_instruction_name(self) -> str:
+        return "frep.outer"
+
+
+@irdl_op_definition
+class FrepInner(FRepOperation):
+    """
+    Repeats the instruction in the body, as if each were in its own body of a for loop,
+    for example:
+
+    ```
+    # Repeat three times, stagger 2, period 2
+    li a0, 3
+    frep.inner a0, 2, 2, 0b0100
+    fadd.d fa0, ft0, ft2
+    fmul.d fa0, ft3, fa0
+    ```
+
+    is equivalent to:
+    ```
+    fadd.d fa0, ft0, ft2
```

Comment:
I'm not sure I fully understand the stagger option yet, but this pattern doesn't seem to be correct?
```suggestion
    fadd.d fa0, ft0, ft3
```

Reply:
I think the mask is for `rs1` only, no?

Reply:
I copied this from the paper verbatim but I think it's actually wrong there. My instinct is to stick with this until we make it work ourselves and update the doc.